plonky2/src/hash/field_merkle_tree.rs

```diff
@@ -0,0 +1,322 @@
+#[cfg(not(feature = "std"))]
+use alloc::vec;
+#[cfg(not(feature = "std"))]
+use alloc::vec::Vec;
+
+use itertools::Itertools;
+
+use crate::hash::hash_types::RichField;
+use crate::hash::merkle_proofs::MerkleProof;
+use crate::hash::merkle_tree::{
+    capacity_up_to_mut, fill_digests_buf, merkle_tree_prove, MerkleCap,
+};
+use crate::plonk::config::{GenericHashOut, Hasher};
+use crate::util::log2_strict;
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct FieldMerkleTree<F: RichField, H: Hasher<F>> {
+    /// The data stored in the Merkle tree leaves.
+    pub leaves: Vec<Vec<Vec<F>>>,
+
+    /// Merkle tree node hashes, analogous to `digests` in `MerkleTree`.
+    pub digests: Vec<H::Hash>,
+
+    /// Represents the roots of the Merkle tree. This allows for using any layer as the root of the tree.
+    pub cap: MerkleCap<F, H>,
+
+    /// Represents the heights at which leaves reside within the tree.
+    pub leaf_heights: Vec<usize>,
+}
+
+impl<F: RichField, H: Hasher<F>> Default for FieldMerkleTree<F, H> {
+    fn default() -> Self {
```

Comment:
Why not just derive Default?  Or is this subtly different?

Reply:
the trait `Default` is not implemented for `<C as GenericConfig<D>>::Hasher`

Reply:
See this pr: https://github.com/0xPolygonZero/plonky2/pull/1219/files

Reply:
> the trait `Default` is not implemented for `<C as GenericConfig<D>>::Hasher`

We don't need that implementation, because you are only using your Hashers inside of Vectors.

It works on my machine, with the derived instance.

Reply:
```diff
diff --git a/plonky2/src/hash/field_merkle_tree.rs b/plonky2/src/hash/field_merkle_tree.rs
index 1512ffa0..f52b2780 100644
--- a/plonky2/src/hash/field_merkle_tree.rs
+++ b/plonky2/src/hash/field_merkle_tree.rs
@@ -13,7 +13,7 @@ use crate::hash::merkle_tree::{
 use crate::plonk::config::{GenericHashOut, Hasher};
 use crate::util::log2_strict;
 
-#[derive(Clone, Debug, Eq, PartialEq)]
+#[derive(Clone, Debug, Default, Eq, PartialEq)]
 pub struct FieldMerkleTree<F: RichField, H: Hasher<F>> {
     /// The data stored in the Merkle tree leaves.
     pub leaves: Vec<Vec<Vec<F>>>,
(1/2) Stage this hunk [y,n,q,a,d,j,J,g,/,e,?]? y
@@ -28,17 +28,6 @@ pub struct FieldMerkleTree<F: RichField, H: Hasher<F>> {
     pub leaf_heights: Vec<usize>,
 }
 
-impl<F: RichField, H: Hasher<F>> Default for FieldMerkleTree<F, H> {
-    fn default() -> Self {
-        Self {
-            leaves: Vec::new(),
-            digests: Vec::new(),
-            cap: MerkleCap::default(),
-            leaf_heights: Vec::new(),
-        }
-    }
-}
-
 impl<F: RichField, H: Hasher<F>> FieldMerkleTree<F, H> {
     /// Each element in the `leaves` vector represents a matrix (a vector of vectors).
     /// The height of each matrix should be a power of two.
```

Reply:
See: https://github.com/0xmozak/plonky2/pull/79#issuecomment-2039291412

Reply:
I for one prefer this slightly longer, but more narrow syntax.

Reply:
Huh, this could have just been derived, no changes necessary.

Do you want to implement all derives by default?